### DISCUSSION TITLE
PHP <=7.3 Opcache is broken on Bookworm

### DISCUSSION THREAD

**Discussion by mauzi** [oerdnj/deb.sury.org#2008] [General]
Hi,

**- Debian 11 (Bullseye) amd64, up-to-date @ today, PHP-FPM:**
7.1.33-54+0\~20230814.83+debian11\~1.gbp8adb1a - everything works flawlessly
7.2.34-40+0\~20230814.86+debian11\~1.gbpe8e12b - everything works flawlessly
7.3.33-12+0\~20230814.109+debian11\~1.gbpa3309d - everything works flawlessly
7.4.33-7+0\~20230814.85+debian11\~1.gbp091944 - everything works flawlessly

**- Debian 12 (Bookworm) amd64, up-to-date @ today, PHP-FPM:**
7.1.33-54+0\~20230815.84+debian12\~1.gbp8adb1a - some Wordpress, Joomla, and other sites fail (100% CPU usage for several minutes, then timeout)
7.2.34-40+0\~20230815.87+debian12\~1.gbpe8e12b - some Wordpress, Joomla, and other sites fail (100% CPU usage several for minutes, then timeout)
7.3.33-12+0\~20230815.111+debian12\~1.gbpa3309d - some Wordpress, Joomla, and other sites fail (100% CPU usage several for minutes, then timeout)
7.4.33-7+0\~20230815.86+debian12\~1.gbp091944 - everything works flawlessly

Do you have any idea?
Thanks!

**Comment by oerdnj** [comment]
Try capturing the stack with `eu-stack`. Alternatively, you might want to do some `perf record` on the php process to sample what's happening.

**Reply by mauzi** [reply]
It has something to do with libgd3.
The problem occurs in functions using the gd extension.
Investigating further...

**Reply by mauzi** [reply]
Please forget my post about libgd3.
It's the opcache.
I have disabled the opcache, and now everything work good. Re-enabling opcache 100% reproduces the problem.

What can be the difference between php7.3-fpm on Debian 11 vs. Debian 12? Are the build options the same?

**Reply by oerdnj** [reply]
Different compiler. The opcache is a bit fragile, so this doesn't surprise me at all.

**Reply by mauzi** [reply]
Look at this code:

```
$x = 0;
$y = hexdec("0200");

for ($i = 0; $i < 100; $i++) {
    if ($x >= $y) {
        $i += $y % 4;
        $x = 0;
    }

    $x++;
}
```

This code always fail on my Debian 12 (amd64) box with PHP 7.3-FPM + opcache enabled.

But:
`var_dump($y)` is int(512), but if I type cast it, then the code does not fail:

```
$y = (int)hexdec("0200");
```

The if() expression should never eval to TRUE, but if I change `$i += $width % 4;` to `$i += 5;` then the code does not fail.

**Comment by oerdnj** [comment]
This one could be worked around with `opcache.optimization_level=0x7FFEBF5F`
